srtcore/core.cpp

```diff
@@ -4678,24 +4779,70 @@ int CUDT::sendmsg(const char* data, int len, int msttl, bool inorder, uint64_t s
     if (bCongestion)
         throw CUDTException(MJ_AGAIN, MN_CONGESTION, 0);
 #endif /* SRT_ENABLE_ECN */
-    return len;
+    return size;
 }
 
-int CUDT::recvmsg(char* data, int len)
+int CUDT::recv(char* data, int len)
 {
-    uint64_t srctime;
-    return(CUDT::recvmsg(data, len, srctime));
+    if (!m_bConnected || !m_Smoother.ready())
+        throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
+
+    if (len <= 0)
+    {
+        LOGC(dlog.Error) << "Length of '" << len << "' supplied to srt_recv.";
+        throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+    }
+
+    if (m_bMessageAPI)
+    {
+        SRT_MSGCTRL mctrl = srt_msgctrl_default;
+        return receiveMessage(data, len, &mctrl);
+    }
+
+    return receiveBuffer(data, len);
 }
 
 int CUDT::recvmsg(char* data, int len, uint64_t& srctime)
 {
+    if (!m_bConnected || !m_Smoother.ready())
+        throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
+
+    if (len <= 0)
+    {
+        LOGC(dlog.Error) << "Length of '" << len << "' supplied to srt_recvmsg.";
+        throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+    }
+
+    if (m_bMessageAPI)
+    {
+        SRT_MSGCTRL mctrl = srt_msgctrl_default;
+        int ret = receiveMessage(data, len, &mctrl);
+        srctime = mctrl.srctime;
+        return ret;
+    }
+
+    return receiveBuffer(data, len);
+}
 
+int CUDT::recvmsg2(char* data, int len, SRT_MSGCTRL* mctrl)
+{
     if (!m_bConnected || !m_Smoother.ready())
         throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
 
     if (len <= 0)
-        return 0;
+    {
+        LOGC(dlog.Error) << "Length of '" << len << "' supplied to srt_recvmsg.";
+        throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+    }
+
+    if (m_bMessageAPI)
+        return receiveMessage(data, len, mctrl);
+
+    return receiveBuffer(data, len);
+}
 
+int CUDT::receiveMessage(char* data, int len, SRT_MSGCTRL* mctrl)
```

Comment:
There ought be a debug assert down this function checking for msgctl != NULL. Alternative is to change msgctl to be a reference.